### PR TITLE
added cache for economic metrics + removed total staked endpoint

### PR DESCRIPTION
**Pull request by bogdan-rosianu**
Added a generic cache for a `GenericApiResponse` and used it to store the economic metrics into it.
Also removed the `total-staked` endpoint

### CODE REVIEW

File: process/cache/genericApiResponseMemCacher.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	if garmc.storedResponse == nil {
+		return nil, ErrNilGenericApiResponseInCache
+	}
+
+	return garmc.storedResponse, nil
+}
+
+// Store will update the generic api response response in cache
+func (garmc *genericApiResponseMemoryCacher) Store(genericApiResponse *data.GenericAPIResponse) error {
+	if genericApiResponse == nil {
```
Comment (iulianpascalau) [inline process/cache/genericApiResponseMemCacher.go:37]: we might drop this condition

File: process/economicMetrics.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+// GetEconomicsDataMetrics will return the economic metrics from cache
+func (nsp *NodeStatusProcessor) GetEconomicsDataMetrics() (*data.GenericAPIResponse, error) {
+	economicMetricsToReturn, err := nsp.economicMetricsCacher.Load()
+	if err == nil {
+		return economicMetricsToReturn, nil
+	}
+
+	log.Info("economic metrics: cannot get from cache. Will fetch from API", "error", err.Error())
+
+	return nsp.getEconomicsDataMetricsFromApi()
```
Comment (iulianpascalau) [inline process/economicMetrics.go:22]: we might remove this extra call + the info on line 20 and directly return error here

File: process/economicMetrics.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+		log.Info("economics data request", "shard id", observer.ShardId, "observer", observer.Address)
+		return responseNetworkMetrics, nil
+	}
+
+	return nil, ErrSendingRequest
+}
+
+// StartCacheUpdate will update the economic metrics cache at a given time
+func (nsp *NodeStatusProcessor) StartCacheUpdate() {
+	go func() {
```
Comment (iulianpascalau) [inline process/economicMetrics.go:53]: we might add a guard variable here: initiated to 0 on L53, incremented on L57, reset on L61. if that value exceeds a certain threshold, store nil in the economicMetricsCacher

File: process/nodeStatusProcessor.go [hunk 4]
```diff
@@ -17,28 +18,36 @@ const NetworkStatusPath = "/network/status"
...
+	economicMetricsCacher GenericApiResponseCacheHandler,
+	cacheValidityDuration time.Duration,
+) (*NodeStatusProcessor, error) {
 	if check.IfNil(processor) {
 		return nil, ErrNilCoreProcessor
 	}
+	if check.IfNil(economicMetricsCacher) {
+		return nil, ErrNilEconomicMetricsCacher
+	}
+	if cacheValidityDuration <= 0 {
```
Comment (iulianpascalau) [inline process/nodeStatusProcessor.go:43]: could have been a larger value like 10 millisecond

File: process/economicMetrics.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+		for {
+			economicMetrics, err := nsp.getEconomicsDataMetricsFromApi()
+			if err != nil {
+				log.Warn("economic metrics: get from API", "error", err.Error())
+			}
+
+			if economicMetrics != nil {
+				err = nsp.economicMetricsCacher.Store(economicMetrics)
+				if err != nil {
+					log.Warn("validator statistics: store in cache", "error", err.Error())
```
Comment (miiu96) [inline process/economicMetrics.go:63]: update log

File: process/economicMetrics_test.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+	"github.com/ElrondNetwork/elrond-proxy-go/process"
+	"github.com/ElrondNetwork/elrond-proxy-go/process/mock"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestNodeStatusProcessor_GetEconomicsDataMetricsShouldReturnDataFromApiBecauseCacheDataIsNil(t *testing.T) {
+	t.Parallel()
+
+	httpWasCalled := false
+	// set nil response in cache
```
Comment (miiu96) [inline process/economicMetrics_test.go:18]: remove comment

File: process/cache/genericApiResponseMemCacher.go (hunk 1 above)
Comment (bogdan-rosianu) [reply process/cache/genericApiResponseMemCacher.go:37]: done

File: process/economicMetrics.go (hunk 2 above)
Comment (bogdan-rosianu) [reply process/economicMetrics.go:22]: done

File: process/economicMetrics.go (hunk 3 above)
Comment (bogdan-rosianu) [reply process/economicMetrics.go:53]: done

File: process/nodeStatusProcessor.go (hunk 4 above)
Comment (bogdan-rosianu) [reply process/nodeStatusProcessor.go:43]: it is ok. it's basically a protection for default toml values

File: process/economicMetrics.go (hunk 5 above)
Comment (bogdan-rosianu) [reply process/economicMetrics.go]: removed that part

File: process/economicMetrics_test.go (hunk 6 above)
Comment (bogdan-rosianu) [reply process/economicMetrics_test.go]: removed that part

**Review by iulianpascalau — APPROVED** [review]

**Review by miiu96 — APPROVED** [review]

**Review by stefanandy — APPROVED** [review]
System Test Passed

**Pull request merged**